sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/SubstituteHintsSuite.scala

```diff
@@ -0,0 +1,208 @@
+      BroadcastHint(testRelation),
+      caseSensitive = false)
+
+    checkAnalysis(
+      Hint("MAPJOIN", Seq("table"), UnresolvedRelation(TableIdentifier("TaBlE"), Some("TbL"))),
+      BroadcastHint(testRelation),
+      caseSensitive = false)
+
+    checkAnalysis(
+      Hint("MAPJOIN", Seq("TaBlE"), UnresolvedRelation(TableIdentifier("TaBlE"), Some("TbL"))),
+      BroadcastHint(testRelation))
+
+    checkAnalysis(
+      Hint("MAPJOIN", Seq("table"), UnresolvedRelation(TableIdentifier("TaBlE"), Some("TbL"))),
+      testRelation)
+  }
+
+  test("single hint") {
+    checkAnalysis(
+      Hint("MAPJOIN", Seq("TaBlE"),
+        Project(Seq(UnresolvedAttribute("a")),
+          UnresolvedRelation(TableIdentifier("TaBlE"), Some("TbL")))),
+      Project(testRelation.output, BroadcastHint(testRelation)))
+
+    checkAnalysis(
+      Hint("MAPJOIN", Seq("TaBlE"),
+        Project(Seq(UnresolvedAttribute("t.a")),
+          Join(
+            UnresolvedRelation(TableIdentifier("TaBlE"), Some("TbL")).as("t"),
+            UnresolvedRelation(TableIdentifier("TaBlE2"), Some("TbL")).as("u"),
+            Inner,
+            None
+          ))),
+      Project(testRelation.output,
+        Join(
+          BroadcastHint(testRelation),
+          testRelation2,
+          Inner,
+          None
+        )))
+
+    checkAnalysis(
+      Hint("MAPJOIN", Seq("TaBlE2"),
+        Project(Seq(UnresolvedAttribute("t.a")),
+          Join(
+            UnresolvedRelation(TableIdentifier("TaBlE"), Some("TbL")).as("t"),
+            UnresolvedRelation(TableIdentifier("TaBlE2"), Some("TbL")).as("u"),
+            Inner,
+            None
+          ))),
+      Project(testRelation.output,
+        Join(
+          testRelation,
+          BroadcastHint(testRelation2),
+          Inner,
+          None
+        )))
+  }
+
+  test("single hint with multiple parameters") {
+    checkAnalysis(
+      Hint("MAPJOIN", Seq("TaBlE", "TaBlE"),
+        Project(Seq(UnresolvedAttribute("t.a")),
+          Join(
+            UnresolvedRelation(TableIdentifier("TaBlE"), Some("TbL")).as("t"),
+            UnresolvedRelation(TableIdentifier("TaBlE2"), Some("TbL")).as("u"),
+            Inner,
+            None
+          ))),
+      Project(testRelation.output,
+        Join(
+          BroadcastHint(testRelation),
+          testRelation2,
+          Inner,
+          None
+        )))
+
+    checkAnalysis(
+      Hint("MAPJOIN", Seq("TaBlE", "TaBlE2"),
+        Project(Seq(UnresolvedAttribute("t.a")),
+          Join(
+            UnresolvedRelation(TableIdentifier("TaBlE"), Some("TbL")).as("t"),
+            UnresolvedRelation(TableIdentifier("TaBlE2"), Some("TbL")).as("u"),
+            Inner,
+            None
+          ))),
+      Project(testRelation.output,
+        Join(
+          BroadcastHint(testRelation),
+          BroadcastHint(testRelation2),
+          Inner,
+          None
+        )))
+  }
+
+  test("duplicated nested hints are transformed into one") {
```

Comment:
can we use the dsl to shorten the test cases? they are pretty long and very difficult to read.